# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webgpu/ComputePipeline.cpp
# Commit: dd10bf41fcea
# Full Hash: dd10bf41fcea6aa2f40d4528f4a7afc22b1f389e
# Author: Nicolas Silva <nsilva@mozilla.com>
# Date: 2024-04-05 21:43:02
# Regressor Bug: 1881518
# File Overlap Count: 1
# Description:
#   Bug 1881518 - Recycle indices again. r=webgpu-reviewers,bradwerth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D202433
# ==============================================================================

diff -r b2cba928a7bb -r dd10bf41fcea dom/webgpu/ComputePipeline.cpp
--- a/dom/webgpu/ComputePipeline.cpp	Fri Apr 05 07:28:07 2024 +0000
+++ b/dom/webgpu/ComputePipeline.cpp	Fri Apr 05 07:42:05 2024 +0000
@@ -27,17 +27,28 @@
 ComputePipeline::~ComputePipeline() { Cleanup(); }
 
 void ComputePipeline::Cleanup() {
-  if (mValid && mParent) {
-    mValid = false;
-    auto bridge = mParent->GetBridge();
-    if (bridge && bridge->IsOpen()) {
-      bridge->SendComputePipelineDrop(mId);
-      if (mImplicitPipelineLayoutId) {
-        bridge->SendImplicitLayoutDrop(mImplicitPipelineLayoutId,
-                                       mImplicitBindGroupLayoutIds);
-      }
+  if (mValid) {
+    return;
+  }
+  mValid = false;
+
+  auto bridge = mParent->GetBridge();
+  if (bridge->IsOpen()) {
+    bridge->SendComputePipelineDrop(mId);
+    if (mImplicitPipelineLayoutId) {
+      bridge->SendImplicitLayoutDrop(mImplicitPipelineLayoutId,
+                                     mImplicitBindGroupLayoutIds);
     }
   }
+
+  if (mImplicitPipelineLayoutId) {
+    wgpu_client_free_pipeline_layout_id(bridge->GetClient(),
+                                        mImplicitPipelineLayoutId);
+  }
+
+  for (const auto& id : mImplicitBindGroupLayoutIds) {
+    wgpu_client_free_bind_group_layout_id(bridge->GetClient(), id);
+  }
 }
 
 already_AddRefed<BindGroupLayout> ComputePipeline::GetBindGroupLayout(